src/components/application_manager/rpc_plugins/vehicle_info_plugin/src/commands/mobile/subscribe_vehicle_data_request.cc

```diff
@@ -0,0 +1,506 @@
+                      "Subscription for VehicleDataType "
+                          << it->first
+                          << " is unsuccessfull. "
+                             "Unsubscribing app with connection key "
+                          << connection_key() << " from it.");
+        app->UnsubscribeFromIVI(it->second);
+      }
+    }
+  }
+}
+
+struct SubscribedToIVIPredicate {
+  int32_t vehicle_info_;
+  SubscribedToIVIPredicate(int32_t vehicle_info)
+      : vehicle_info_(vehicle_info) {}
+  bool operator()(const ApplicationSharedPtr app) const {
+    return app ? app->IsSubscribedToIVI(vehicle_info_) : false;
+  }
+};
+
+bool SubscribeVehicleDataRequest::IsSomeoneSubscribedFor(
+    const uint32_t param_id) const {
+  LOG4CXX_AUTO_TRACE(logger_);
+  SubscribedToIVIPredicate finder(param_id);
+  DataAccessor<ApplicationSet> accessor = application_manager_.applications();
+  ApplicationSetConstIt it = std::find_if(
+      accessor.GetData().begin(), accessor.GetData().end(), finder);
+  return it != accessor.GetData().end();
+}
+
+void SubscribeVehicleDataRequest::CheckVISubscribtions(
+    ApplicationSharedPtr app,
+    std::string& out_info,
+    mobile_apis::Result::eType& out_result_code,
+    smart_objects::SmartObject& out_response_params,
+    smart_objects::SmartObject& out_request_params,
+    bool& out_result) {
+  CheckVISubscriptions(app,
+                       out_info,
+                       out_result_code,
+                       out_response_params,
+                       out_request_params,
+                       out_result);
+}
+
+void SubscribeVehicleDataRequest::CheckVISubscriptions(
+    ApplicationSharedPtr app,
+    std::string& out_info,
+    mobile_apis::Result::eType& out_result_code,
+    smart_objects::SmartObject& out_response_params,
+    smart_objects::SmartObject& out_request_params,
+    bool& out_result) {
+  // counter for items to subscribe
+  app_mngr::VehicleInfoSubscriptions::size_type items_to_subscribe = 0;
+  // counter for subscribed items by application
+  uint32_t subscribed_items = 0;
+
+  const VehicleData& vehicle_data = MessageHelper::vehicle_data();
+  VehicleData::const_iterator it = vehicle_data.begin();
+
+  HmiInterfaces::InterfaceState interface_state =
+      application_manager_.hmi_interfaces().GetInterfaceState(
+          HmiInterfaces::HMI_INTERFACE_VehicleInfo);
+
+  const bool is_interface_not_available =
+      interface_state == HmiInterfaces::STATE_NOT_AVAILABLE;
+
+  for (; vehicle_data.end() != it; ++it) {
+    const std::string& key_name = it->first;
+    if ((*message_)[strings::msg_params].keyExists(key_name)) {
+      const bool is_key_enabled =
+          (*message_)[strings::msg_params][key_name].asBool();
+      if (is_key_enabled) {
+        ++items_to_subscribe;
+      }
+      if (!is_interface_not_available && is_key_enabled) {
+        mobile_apis::VehicleDataType::eType key_type = it->second;
+        if (app->IsSubscribedToIVI(key_type)) {
```

Comment:
IsSubscribedToIVI() shall be a function moved into vehicle info plugin

Reply:
@yang1070 fixed 0afeef687727606c83f2627971a614f671005bf4